### PR TITLE
Add Telegram alerting

### LINKED ISSUE #36: [FEATURE] Telegram notification alerts

### PR DESCRIPTION
**Pull request by zufardhiyaulhaq**
### Summary

- Add telegram alerting with markdown templating (issue https://github.com/kitabisa/teler/issues/36)
- not support silent mode

### Proposed changes

This PR fixes/implements the following **bugs/features**:

- Add telegram alerting

### How has this been tested?

configuring teler.yaml using telegram alert provider

Proof:

example logging in telegram
```
######################
Warning: teler Alert
######################

 Crawler

Request:  / HTTP/1.1
Date:  +0700
IP Address: ::1
User Agent: curl/7.47.0
Referrer: -
Status code: 200
Bytes sent: 5

######################
Warning: teler Alert
######################

 Bruteforce

Request:  /_private HTTP/1.1
Date:  +0700
IP Address: 127.0.0.1
User Agent:  (Windows NT 6.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/28.0.1468.0 Safari/537.36
Referrer: -
Status code: 404
Bytes sent: 580

######################
Warning: teler Alert
######################

 Web Attack: Detects JavaScript location/document property access and window access obfuscation

Request:  /?_=1%20AND%20%28SELECT%202000%20FROM%28SELECT%20COUNT%28%2A%29%2CCONCAT%280x716b786b71%2C%28SELECT%20%28ELT%282000%3D2000%2C1%29%29%29%2C0x716a707871%2CFLOOR%28RAND%280%29%2A2%29%29x%20FROM%20INFORMATION_SCHEMA.PLUGINS%20GROUP%20BY%20x%29a%29 HTTP/1.1
Date:  +0700
IP Address: ::1
User Agent:  (http://sqlmap.org)
Referrer: -
Status code: 200
Bytes sent: 160
```

### Closing issues

Fixes #36

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly.
- [x] I have followed the guidelines in our [CONTRIBUTING.md](https://github.com/kitabisa/teler/blob/development/.github/CONTRIBUTING.md) document.
- [ ] I have written new tests for my changes.
  - [ ] My changes successfully ran and pass tests locally.

### CODE REVIEW

**Comment by dwisiswant0** [comment]
Can you please provide a screenshot attachment of alert, @zufardhiyaulhaq?

**Comment by zufardhiyaulhaq** [comment]
@dwisiswant0 give you the screenshot
gambar

**Comment by dwisiswant0** [comment]
It looks like the first word of threat category got trimmed, e.g. ~Bad~ Crawler, ~Common~ Web Attack.

**Comment by zufardhiyaulhaq** [comment]
let me check, that is weird since the markdown generated from the template is normal.
```
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#
*Warning: teler Alert*
\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#

```Bad Crawler```

*Request*: ```GET / HTTP/1.1```
*Date*: ```28/Oct/2020:08:23:43 +0700```
*IP Address*: ```127.0.0.1```
*User Agent*: ```check_http/v2.1.2 (monitoring-plugins 2.1.2)```
*Referrer*: ```-```
*Status code*: ```200```
*Bytes sent*: ```5```
```

**Comment by zufardhiyaulhaq** [comment]
@dwisiswant0 this is done.
gambar

File: README.md [hunk 1]
```diff
@@ -319,6 +319,10 @@ notifications:
     token: "xoxb-..."
     color: "#ffd21a"
     channel: "G30SPKI"
+  telegram:
```
Comment (dwisiswant0) [inline README.md:322]: Please trailing newline.

File: teler.example.yaml [hunk 2]
```diff
@@ -23,6 +23,7 @@ rules:
 
 alert:
   active: false
+  # supported `slack` or `telegram`
```
Comment (dwisiswant0) [inline teler.example.yaml:26]: Please remove this comment.

**Review by dwisiswant0 — CHANGES_REQUESTED** [review]
Hi @zufardhiyaulhaq,
After reviewing, I still need some revisions of this.

Please note that we provide the `parse_mode` option in the configuration file because we have to provide some formatting options for the user. If I see; you only provide `MarkdownV2` when referring to currently templates you created, but not for ~`Markdown` and~ `HTML`.

References:
- https://core.telegram.org/bots/api#formatting-options

File: internal/alert/telegram.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package alert
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+	"text/template"
+	"time"
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go:8]: you cannot remove this, this is required
```
internal/alert/telegram.go:9:13: undefined: strconv
internal/alert/telegram.go:25:3: undefined: time
internal/alert/telegram.go:33:13: undefined: bytes
internal/alert/telegram.go:35:19: undefined: template
```

**Comment by zufardhiyaulhaq** [comment]
@dwisiswant0 I am aware that we can use MarkdownV2 or HTML, but why in the first place we can make user can customize this? since we manage the template and I believe it is not necessary for the user to select HTML or MarkdownV2 if the user cannot customize the template from the first place.

File: internal/alert/telegram.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
+package alert
+
+import (
+	"bytes"
+	"fmt"
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go:5]: needed by the code
```
internal/alert/telegram.go:9:13: undefined: strconv
internal/alert/telegram.go:25:3: undefined: time
internal/alert/telegram.go:33:13: undefined: bytes
internal/alert/telegram.go:35:19: undefined: template
```

File: internal/alert/telegram.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"strconv"
+	"text/template"
+	"time"
+
+	telegramBot "github.com/go-telegram-bot-api/telegram-bot-api"
+	"github.com/projectdiscovery/gologger"
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go]: this is done

File: internal/alert/telegram.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+)
+
+func toTelegram(token string, chatID string, parseMode string, log map[string]string) {
+	id, err := strconv.ParseInt(chatID, 10, 64)
+	if err != nil {
+		gologger.Errorf(err.Error())
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go]: this is done

File: .gitignore [hunk 7]
```diff
@@ -1,4 +1,5 @@
 bin/*
 teler.yaml
+example.log
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply .gitignore]: this is done

File: internal/alert/telegram.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	api, err := telegramBot.NewBotAPI(token)
+	if err != nil {
+		gologger.Errorf(err.Error())
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go]: this is done

File: internal/alert/telegram.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	var buffer bytes.Buffer
+
+	template, err := template.ParseFiles(fmt.Sprintf("internal/alert/template/telegram.tmpl"))
+	if err != nil {
+		gologger.Warningf(err.Error())
+	}
+
+	err = template.Execute(&buffer, log)
+	if err != nil {
+		gologger.Warningf(err.Error())
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go]: this is done

File: internal/alert/telegram.go [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+	message := telegramBot.NewMessage(id, telegramMessage(log))
+	message.ParseMode = parseMode
+
+	for {
+		_, err = api.Send(message)
+		time.Sleep(1 * time.Second)
+		if err == nil {
+			break
+		}
+	}
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go:30]: this is done

File: internal/alert/telegram.go [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+func telegramMessage(log map[string]string) string {
+	var buffer bytes.Buffer
+
+	template, err := template.ParseFiles(fmt.Sprintf("internal/alert/template/telegram.tmpl"))
```
[the remark this answers is not in the document]
Comment (zufardhiyaulhaq) [reply internal/alert/telegram.go]: this is done

File: README.md (hunk 1 above)
Comment (zufardhiyaulhaq) [reply README.md:323]: this is done

File: teler.example.yaml (hunk 2 above)
Comment (zufardhiyaulhaq) [reply teler.example.yaml]: this is done

File: internal/alert/telegram.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dwisiswant0) [reply internal/alert/telegram.go]: I mean `time` package. Mark as resolved.

File: internal/alert/telegram.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dwisiswant0) [reply internal/alert/telegram.go]: I mean `time` package. Mark as resolved.

**Comment by dwisiswant0** [comment]
What I mean is, when the user choose HTML; the template format we'll use is to generate a message using table fields, since the `Markdown(V2)` style doesn't support generating tables for messages on Telegram.

I'm going to change the base branch to `development`, as I will refactor configuration requirements and so on from this. You makes a good point, @zufardhiyaulhaq!

Thanks for your contributions! Approving now.

**Review by dwisiswant0 — APPROVED** [review]

**Comment by dwisiswant0** [comment]
@all-contributors please add @zufardhiyaulhaq for code.

**Pull request merged**